docs/decisions/0004-managing-django-settings.rst

```diff
@@ -0,0 +1,37 @@
+Status
+======
+
+Accepted
+
+
+Context
+=======
+
+Settings are confusing right now and there are way too many layers. We have
+multiple python settings file that "inherit" from each other. In some cases some
+layers of the inheritance tree also pull settings from config files leading to
+more confusion.
+
+
+Decision
+========
+Rather than having multiple python settings files that override things
+differently we will move to just a few python files with most of the settings
+variance living in YAML config files that are managed by environment operators.
+
+common.py - This file will house the defaults for all settings that are
+referenced in the default installation of the edx-platform.
+
+production.py - This settings file will pull the defaults from common.py and
+then override them with settings pulled from a single YAML config file. For the
+most part this will be a wholesale replacement for any complex values(dicts,
+lists, etc) but for some specific settings they will be additive for now.
+
+eg. ADDITIONAL_INSTALLED_APPS, or ADDITIONAL_MIDDLEWARE
```

Comment:
I think dicts and list are different animals:  
- Lists need `ADDITIONAL_` settings so you can choose between adding or replacing the defaults.
- Dicts are still just key/value pairs that allow someone to group settings.  I would love it if updating were the rule and overwriting were the exception for dicts.  But, I can see arguments against this, so I am simply stating my preference.

Reply:
I'd rather not see dict merging at all.  It results in a final settings dict that may not be visible anywhere except in production making it hard to debug and understand the current state of what's set.  Ansible did this and it caused more problems than it solved in my opinion.  I'd rather move towards things being more explicit and providing entrypoints for adding behavior to the core.

Reply:
👍 
1. Sounds like you would have the same argument for lists (i.e. avoiding `ADDITIONAL_` settings), as well, but not sure?
2. I think this is good content for the ADR, whether it is next to the decision or in rejected alternatives, or where ever.  It helps clarify and explain your decision, even if cleaning up certain exceptions doesn't come until later.